syntax = "proto3";

package echo;

service EchoService {
  rpc LoginUser (LoginUserRequest) returns (LoginUserReply) {}
  rpc RegisterUser (RegisterUserRequest) returns (RegisterUserReply) {}
  rpc AddTickerUtente(AddTickerUtenteRequest) returns (AddTickerUtenteReply){}
  rpc ShowTickersUser(ShowTickersUserRequest) returns (ShowTickersUserReply){}
  rpc DeleteTickerUser(DeleteTickerUserRequest) returns (DeleteTickerUserReply){}
  rpc UpdateUser (UpdateUserRequest) returns (UpdateUserReply) {}
  rpc DeleteUser (DeleteUserRequest) returns (DeleteUserReply) {}
  rpc GetLatestValue (GetLatestValueRequest) returns (GetLatestValueReply) {}
  rpc GetAverageValue (GetAverageValueRequest) returns (GetAverageValueReply) {}
  rpc UpdateMinMaxValue (UpdateMinMaxValueRequest) returns (UpdateMinMaxValueReply) {}

}

message LoginUserRequest {
  string email = 1;
}

message LoginUserReply {
  bool success = 1;
  string message = 2;
}

message RegisterUserRequest {
  string email = 1;
  string ticker = 2; // Codice dell'azione (es. "AAPL" per Apple)
  float max_value = 3; // Valore massimo (numero decimale)
  float min_value = 4; // Valore minimo (numero decimale)
}

message RegisterUserReply {
  bool success = 1; // Indica se la registrazione Ã¨ avvenuta con successo
  string message = 2; // Messaggio di conferma o errore
}


message AddTickerUtenteRequest {
  string ticker = 1;
  float max_value = 2; // Valore massimo (numero decimale)
  float min_value = 3; // Valore minimo (numero decimale)
}

message AddTickerUtenteReply {
  bool success = 1;
  string message = 2;
}

message ShowTickersUserRequest {

}

message ShowTickersUserReply {
  bool success = 1;
  string message = 2;
  string ticker = 3;
}

message DeleteTickerUserRequest {
  string ticker = 1;
}

message DeleteTickerUserReply {
  bool success = 1;
  string message = 2;
}


message UpdateUserRequest {
  string ticker_old = 1;
  string ticker = 2; // Nuovo codice dell'azione
  float max_value = 3; // Valore massimo (numero decimale)
  float min_value = 4; // Valore minimo (numero decimale)
}

message UpdateUserReply {
  bool success = 1;
  string message = 2;
}


message UpdateMinMaxValueRequest {
    string ticker = 1;
    float max_value = 2;
    float min_value = 3;
}

message UpdateMinMaxValueReply {
    bool success = 1;
    string message = 2;
    string ticker = 3;
}


message DeleteUserRequest {
}

message DeleteUserReply {
  bool success = 1;
  string message = 2;
}


message GetLatestValueRequest {
  string ticker = 1;
}

message GetLatestValueReply {
  bool success = 1;
  string message = 2;
  string ticker = 3;
  double stock_value = 4;
  string timestamp = 5;
}

message GetAverageValueRequest {
  string ticker = 1;
  int32 num_values = 2;
}

message GetAverageValueReply {
  bool success = 1;
  string message = 2;
  double average_stock_value = 3;
  string timestamp = 4;
  string ticker = 5;
}
